File: development/src/replydata.c
//
// --------------------------------------------------------------------------
//  Gurux Ltd
//
//
//
// Filename:        $HeadURL$
//
// Version:         $Revision$,
//                  $Date$
//                  $Author$
//
// Copyright (c) Gurux Ltd
//
//---------------------------------------------------------------------------
//
//  DESCRIPTION
//
// This file is a part of Gurux Device Framework.
//
// Gurux Device Framework is Open Source software; you can redistribute it
// and/or modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; version 2 of the License.
// Gurux Device Framework is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU General Public License for more details.
//
// This code is licensed under the GNU General Public License v2.
// Full text may be retrieved at http://www.gnu.org/licenses/gpl-2.0.txt
//---------------------------------------------------------------------------

#include <string.h> /* memset */
#include "../include/gxmem.h"
#if _MSC_VER > 1400
#include <crtdbg.h>
#endif
#include "../include/replydata.h"

/**
* @return Is more data available.
*/
unsigned char reply_isMoreData(gxReplyData* reply)
{
    return reply->moreData != DLMS_DATA_REQUEST_TYPES_NONE;
}

/**
* Initialize reply structure.
*/
void reply_init(gxReplyData* reply)
{
    reply->invokeId = 0;
    reply->commandType = 0;
    reply->moreData = DLMS_DATA_REQUEST_TYPES_NONE;
    reply->encryptedCommand = reply->command = DLMS_COMMAND_NONE;
    BYTE_BUFFER_INIT(&reply->data);
    reply->complete = 0;
    var_init(&reply->dataValue);
    reply->totalCount = 0;
    reply->readPosition = 0;
    reply->packetLength = 0;
    reply->peek = 0;
    reply->ignoreValue = 0;
    reply->dataType = DLMS_DATA_TYPE_NONE;
    reply->cipherIndex = 0;
#ifdef DLMS_USE_EPOCH_TIME
    reply->time = 0;
#else
    memset(&reply->time, 0, sizeof(struct tm));
#endif // DLMS_USE_EPOCH_TIME
    reply->preEstablished = 0;
    reply->blockNumber = 0;
    reply->blockNumberAck = 0;
    reply->streaming = 0;
    reply->windowSize = 0;
    reply->serverAddress = 0;
    reply->clientAddress = 0;
}

void reply_clear2(gxReplyData* reply, unsigned char clearData)
{
    reply->invokeId = 0;
    reply->moreData = DLMS_DATA_REQUEST_TYPES_NONE;
    reply->encryptedCommand = reply->command = DLMS_COMMAND_NONE;
    if (clearData)
    {
        bb_clear(&reply->data);
        reply->preEstablished = 0;
    }
    reply->complete = 0;
    var_clear(&reply->dataValue);
    reply->totalCount = 0;
    reply->readPosition = 0;
    reply->packetLength = 0;
    reply->peek = 0;
    reply->ignoreValue = 0;
    reply->dataType = DLMS_DATA_TYPE_NONE;
    reply->cipherIndex = 0;
#ifdef DLMS_USE_EPOCH_TIME
    reply->time = 0;
#else
    memset(&reply->time, 0, sizeof(struct tm));
#endif // DLMS_USE_EPOCH_TIME
    reply->blockNumber = 0;
    reply->blockNumberAck = 0;
    reply->streaming = 0;
    reply->windowSize = 0;
    reply->serverAddress = 0;
    reply->clientAddress = 0;
}

void reply_clear(gxReplyData* reply)
{
    reply_clear2(reply, 1);
}

File: development/dlms_cosem/src/reply_data.rs
//! DLMS/COSEM reply data.

use crate::types::Command;
use crate::variant::Variant;

/// DLMS/COSEM reply data.
pub struct ReplyData {
    /// DLMS command.
    pub command: Command,
    /// Encrypted DLMS command.
    pub encrypted_command: Command,
    /// Command type.
    pub command_type: u8,
    /// Reply data.
    pub data: Vec<u8>,
    /// Is reply data complete.
    pub complete: bool,
    /// Reply data value.
    pub data_value: Variant,
    /// Total count of items.
    pub total_count: u16,
    /// Current read position.
    pub read_position: usize,
    /// Packet length.
    pub packet_length: usize,
    /// Is peeking.
    pub peek: bool,
    /// Is ignoring value.
    pub ignore_value: bool,
    /// Data type.
    pub data_type: u8, // Placeholder for DLMS_DATA_TYPE
    /// Cipher index.
    pub cipher_index: u16,
    /// Time.
    pub time: u32, // Placeholder for struct tm
    /// Is pre-established.
    pub pre_established: bool,
    /// Invoke ID.
    pub invoke_id: u8,
    /// Block number.
    pub block_number: u16,
    /// Block number ack.
    pub block_number_ack: u16,
    /// Is streaming.
    pub streaming: bool,
    /// Window size.
    pub window_size: u32,
    /// Server address.
    pub server_address: u16,
    /// Client address.
    pub client_address: u16,
}

impl ReplyData {
    /// Creates a new `ReplyData`.
    pub fn new() -> Self {
        Self {
            command: Command::None,
            encrypted_command: Command::None,
            command_type: 0,
            data: Vec::new(),
            complete: false,
            data_value: Variant::new(),
            total_count: 0,
            read_position: 0,
            packet_length: 0,
            peek: false,
            ignore_value: false,
            data_type: 0,
            cipher_index: 0,
            time: 0,
            pre_established: false,
            invoke_id: 0,
            block_number: 0,
            block_number_ack: 0,
            streaming: false,
            window_size: 0,
            server_address: 0,
            client_address: 0,
        }
    }

    /// Clears the `ReplyData`.
    pub fn clear(&mut self) {
        self.clear2(true);
    }

    /// Clears the `ReplyData`.
    pub fn clear2(&mut self, clear_data: bool) {
        if clear_data {
            self.data.clear();
        }
        self.command = Command::None;
        self.encrypted_command = Command::None;
        self.command_type = 0;
        self.complete = false;
        self.data_value.clear();
        self.total_count = 0;
        self.read_position = 0;
        self.packet_length = 0;
        self.peek = false;
        self.ignore_value = false;
        self.data_type = 0;
        self.cipher_index = 0;
        self.time = 0;
        self.pre_established = false;
        self.invoke_id = 0;
        self.block_number = 0;
        self.block_number_ack = 0;
        self.streaming = false;
        self.window_size = 0;
        self.server_address = 0;
        self.client_address = 0;
    }
}
